tests/conformance/test_conformance.py

```diff
@@ -0,0 +1,449 @@
+    else:
+        bucket.rename_blob(object, new_name)
+
+
+def rewrite_blob(client, _preconditions, bucket, object):
+    new_blob = client.bucket(bucket.name).blob(uuid.uuid4().hex)
+    new_blob.metadata = {"foo": "bar"}
+    if _preconditions:
+        new_blob.rewrite(object, if_generation_match=0)
+    else:
+        new_blob.rewrite(object)
+
+
+def compose_blob(client, _preconditions, bucket, object):
+    blob = client.bucket(bucket.name).blob(object.name)
+    blob_2 = bucket.blob(uuid.uuid4().hex)
+    blob_2.upload_from_string("foo")
+    sources = [blob_2]
+
+    if _preconditions:
+        blob.compose(sources, if_generation_match=object.generation)
+    else:
+        blob.compose(sources)
+
+
+# Method invocation mapping. Methods to retry. This is a map whose keys are a string describing a standard
+# API call (e.g. storage.objects.get) and values are a list of functions which
+# wrap library methods that implement these calls. There may be multiple values
+# because multiple library methods may use the same call (e.g. get could be a
+# read or just a metadata get).
+method_mapping = {
+    "storage.buckets.delete": [delete_bucket],  # S1 start
+    "storage.buckets.get": [get_bucket, reload_bucket],
+    "storage.buckets.getIamPolicy": [get_iam_policy],
+    "storage.buckets.insert": [create_bucket],
+    "storage.buckets.list": [list_buckets],
+    "storage.buckets.lockRententionPolicy": [],  # lock_retention_policy
+    "storage.buckets.testIamPermission": [get_iam_permissions],
+    "storage.notifications.delete": [delete_notification],
+    "storage.notifications.get": [get_notification],
+    "storage.notifications.list": [list_notifications],
+    "storage.objects.get": [get_blob],
+    "storage.objects.list": [list_blobs],  # S1 end
+    "storage.buckets.patch": [patch_bucket],  # S2 start
+    "storage.buckets.setIamPolicy": [],
+    "storage.buckets.update": [update_bucket],
+    "storage.objects.compose": [],  # compose_blob
+    "storage.objects.copy": [copy_blob, rename_blob],
+    "storage.objects.delete": [delete_blob],
+    "storage.objects.insert": [upload_from_string],
+    "storage.objects.patch": [patch_blob],
+    "storage.objects.rewrite": [rewrite_blob],
+    "storage.objects.update": [update_blob],  # S2 end
+    "storage.notifications.insert": [create_notification],  # S4
+}
+
+########################################################################################################################################
+### Helper Methods for Populating Resources ############################################################################################
+########################################################################################################################################
+
+
+def _populate_resource_bucket(client, resources):
+    bucket = client.bucket(uuid.uuid4().hex)
+    client.create_bucket(bucket)
+    resources["bucket"] = bucket
```

Comment:
How are the created resources being cleaned up?  The cleanup needs to be resilient against unexpected failures, both during test runs and while tearing down other resources.

Reply:
These tests are run against an emulator; when it shuts down at the end of the test session, all created resources disappear.